server/src/main/java/org/opensearch/monitor/fs/FsHealthService.java

```diff
@@ -170,14 +172,22 @@ public StatusInfo getHealth() {
 
         @Override
         public void run() {
+            boolean checkEnabled = enabled;
             try {
-                if (enabled) {
-                    setLastRunStartTimeMillis();
-                    monitorFSHealth();
-                    logger.debug("health check succeeded");
+                if (checkEnabled) {
+                    if (checkInProgress.compareAndSet(false, true)) {
+                        setLastRunStartTimeMillis();
```

Comment:
Unless these two operation are synchronised, there is still a chance that ``getHealth()`` call can see a current run in progress, while a stale ``lastRunStartTimeMillis`` value, which can easily breach the threshold. 

Still feel solving this with a single Atomic long value with reset, as in suggestion https://github.com/opensearch-project/OpenSearch/pull/1167#discussion_r699498772 is simplest. Thoughts?

Reply:
it's as simple as just changing the order of two commands.. just missed

Reply:
How about this guys:

```    
  private final AtomicReference<Tuple<Long, Long>> startStopRunTimeMillis = 
        new AtomicReference<>(new Tuple<>(Long.MIN_VALUE, Long.MIN_VALUE));


     @Override
    public StatusInfo getHealth() {
        final Tuple<Long, Long> startStop = startStopRunTimeMillis.get();
        final long startTime = startStop.v1();
        final long stopTime = (startStop.v2() >= 0) ? startStop.v2() : currentTimeMillisSupplier.getAsLong();
        
        StatusInfo statusInfo;
        Set<Path> unhealthyPaths = this.unhealthyPaths;
        if (enabled == false) {
           ...
        } else if (startTime >= 0 && stopTime - startTime > healthyTimeoutThreshold.millis()) {
            statusInfo = new StatusInfo(UNHEALTHY, "healthy threshold breached");
        } 
        ... 
    }

   @Override
    public void run() {
        try {
            if (enabled) {
                setLastRunStartTimeMillis();
                monitorFSHealth();
                logger.debug("health check succeeded");
                setLastRunStopTimeMillis();
            }
        } catch (Exception e) {
            logger.error("health check failed", e);
        }
    }

   private void setLastRunStartTimeMillis() {
        startStopRunTimeMillis.getAndUpdate(l -> 
            new Tuple<>(Math.max(l.v1(), currentTimeMillisSupplier.getAsLong()), Long.MIN_VALUE));
    }
    
    private void setLastRunStopTimeMillis() {
        startStopRunTimeMillis.getAndUpdate(l -> 
            new Tuple<>(l.v1(), Math.max(l.v1(), currentTimeMillisSupplier.getAsLong())));
    }
```

Reply:
This looks good to me. However it looks like we don't need strict synchronization if we check in progress. I'd prefer the current PR approach for simplicity, however open to change, let me know your thoughts?

Reply:
Thanks for the suggestion @reta. This works well. Wondering if we really need to update ``stopTime`` to ``Long.MIN_VALUE`` here since we anyways have discrete values for ``startTime`` and ``stopTime``. Possibly just the conditional check of ``startTime`` > ``stopTime`` helps detect there is a current run is in progress. Having said that, we can also get rid of tuple updates as just independent updates on these two Atomic variables at Start and Stop should siffice.
Let me know what you think.


> it's as simple as just changing the order of two commands.. just missed

@Bukhtawar Agree updating the order, where we update the timestamp first, and then mark the run in progress will solve this.